### ISSUE TITLE
ASTableView dynamic row heights with ASLayout

### ISSUE THREAD

**Issue by robmontesinos** [facebook/AsyncDisplayKit#759]
I am unable to have dynamic heights in an ASTableView using ASLayout - I was able to do it manually using -(CGSize)calculateSizeThatFits:(CGSize)constrainedSize - would you mind checking my code below to see what is going wrong? All the rows are the same height.

Note: I had to manually set captionTextNode frame in -(void)layout which makes it a little inaccurate compared to where ASLayoutSpec places the xPosition - captionSize is calculated in ViewController based on text content

```ObjectiveC
- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize {
    self.iconImageNode.preferredFrameSize = CGSizeMake(kSponsorIconSize, kSponsorIconSize);
    self.titleTextNode.preferredFrameSize = CGSizeMake(_captionSize.width, kTitleHeight);
    self.captionTextNode.preferredFrameSize = _captionSize;
    
    ASStackLayoutSpec *verticalStack = [[ASStackLayoutSpec alloc] init];
    verticalStack.direction = ASStackLayoutDirectionVertical;
    [verticalStack setChildren:@[self.titleTextNode, self.captionTextNode]];
    
    ASStackLayoutSpec *horizontalStack = [[ASStackLayoutSpec alloc] init];
    horizontalStack.direction = ASStackLayoutDirectionHorizontal;
    horizontalStack.spacing = kInnerPadding;
    [horizontalStack setChildren:@[self.iconImageNode, verticalStack]];
    
    ASInsetLayoutSpec *insetSpec = [[ASInsetLayoutSpec alloc] init];
    insetSpec.insets = UIEdgeInsetsMake(kOuterPadding, kOuterPadding, kOuterPadding, kOuterPadding);
    insetSpec.child = horizontalStack;
    
    return insetSpec;
}

- (void)layout {
    [super layout];
    // Manually layout the divider.
    CGFloat pixelHeight = 1.0f / [[UIScreen mainScreen] scale];
    self.divider.frame = CGRectMake(0.0f, 0.0f, self.calculatedSize.width, pixelHeight);
    
    self.captionTextNode.frame = CGRectMake(kTextNodeXPos, kTitleHeight + kOuterPadding, _captionSize.width, _captionSize.height);
    
    NSLog(@"%@\n%@", NSStringFromCGRect(self.frame), NSStringFromCGRect(self.captionTextNode.frame));
}

![Screenshot](https://dl.dropboxusercontent.com/u/47066165/ASDK-screenshot.png)

**Comment by appleguy** [comment]
@robmontesinos cool, thanks for the code.

- Is this in an ASCellNode subclass?
- Why does _captionSize exist?  This is suspect, and indicates the layout system is probably not being used 100% correctly.  It may not be the cause of the issue, but could be related.

**Comment by robmontesinos** [comment]
Yes this is ASCellNode subclass.

_captionSize is calculated in the view controller for the size of the caption ASTextNode

```ObjectiveC
        _captionTextNode = [[ASTextNode alloc] init];
        _captionTextNode.layerBacked = YES;
        _captionTextNode.truncationMode = NSLineBreakByTruncatingTail;
        _captionTextNode.attributedString = [[NSAttributedString alloc] initWithString:sponsor.caption
                                                                            attributes:captionAttributes];

**Comment by appleguy** [comment]
@nguyenhuy based on the screenshot image link (https://dl.dropboxusercontent.com/u/47066165/ASDK-screenshot.png) can you see what's going wrong?  @robmontesinos thanks for including the screenshot, as it is very helpful to understand the use case (BTW, I approve of the app content ;) )

I'm thinking that you may need to set a property like flexBasis or flexGrow in order to have the horizontal stack increase its height to the full height the vertical stack would use.  However this doesn't seem like it should be absolutely necessary to me in this scenario so I'm looking forward to figuring out the right answer.

**Comment by robmontesinos** [comment]
@appleguy HAHA o app content - test entries at the moment but in the future???

**Comment by robmontesinos** [comment]
If I comment out the caption frame in 'layout' method - I get a frame that is wider than the screen and only one line high, e.g. {{69, 21}, {1020.5, 14.5}}

```ObjectiveC
//    self.captionTextNode.frame = CGRectMake(kTextNodeXPos, kTitleHeight + kOuterPadding, _captionSize.width, _captionSize.height);

**Comment by appleguy** [comment]
@robmontesinos note the code snippet you showed doesn't actually show where _captionSize is calculated.  Wherever you're calculating it, it's almost certainly creating a material main thread cost that will affect performance.  I haven't yet run into a situation where it's necessary to do this manually with ASDK; one of the big advantages of the measurement system is that you can use .calculatedSize (or .calculatedLayout) once measurement is done, which is always complete by the time the -layout method is called.

**Comment by robmontesinos** [comment]
I'll give that a try - it's calculated in nodeForRowAtIndexPath

**Comment by appleguy** [comment]
@robmontesinos ok, that makes sense then.  I totally understand what's happening now.

The manual setting of the frame is causing a mismatch between measurement and sizing:
- During measurement, the system is assuming you want a single line of text because nothing is telling it otherwise...this may be a bug, we need to investigate the right fix.
- During drawing (which happens after the -layout call, which itself happens after measurement), the system sees that you've given it a specific frame size and then tries to draw the text in the size you've provided.  This results in line wrapping, which was not actually part of the layout calculation.

This should be easy / elegant to fix, but because I'm very new to box model layout myself, I need to think about what the correct solution is.

**Comment by robmontesinos** [comment]
Cool beans @appleguy - I'm gonna kill the contentSize and see what happens

**Comment by appleguy** [comment]
Doing that alone won't fix it BTW...you could definitely fix it by using an ASStaticLayoutSpec to contain the text node and "bound" it so that it knows to wrap...  Still think that isn't the best solution though.

**Comment by robmontesinos** [comment]
I removed _captionSize and set flexGrow to YES and got a height of a single line - image below
![single line](https://dl.dropboxusercontent.com/u/47066165/ASDK-screenshot2.png)

**Comment by appleguy** [comment]
@robmontesinos cool, that is expected, and a good sign actually.

Try setting .flexShrink = YES on the vertical stack spec itself and see what happens...

**Comment by appleguy** [comment]
@robmontesinos silly me, I forgot that the Kittens app has a /very/ similar layout.  Sure enough, the key property is .flexShrink.  Let me know when you can confirm that fixes your issue.

This clearly needs to be covered well in documentation / guides.  It may even be reasonable to set it as default-on for textNode.  @rcancro, @nguyenhuy what do you think about that as a default?

Here's the KittenNode implementation:

- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize
{
  _imageNode.preferredFrameSize = _isImageEnlarged ? CGSizeMake(2.0 * kImageSize, 2.0 * kImageSize) : CGSizeMake(kImageSize, kImageSize);
  _textNode.flexShrink = YES;
  
  ASStackLayoutSpec *stackSpec = [[ASStackLayoutSpec alloc] init];
  stackSpec.direction = ASStackLayoutDirectionHorizontal;
  stackSpec.spacing = kInnerPadding;
  [stackSpec setChildren:!_swappedTextAndImage ? @[_imageNode, _textNode] : @[_textNode, _imageNode]];
  
  ASInsetLayoutSpec *insetSpec = [[ASInsetLayoutSpec alloc] init];
  insetSpec.insets = UIEdgeInsetsMake(kOuterPadding, kOuterPadding, kOuterPadding, kOuterPadding);
  insetSpec.child = stackSpec;
  
  return insetSpec;
}

**Comment by robmontesinos** [comment]
I am getting variable width as opposed to variable heights for the caption - the frame is staying the same height 63.0

(-[SponsorCellNode layout]) (SponsorCellNode.m:189) Node {{0, 0}, {320, 63}}
Caption {{69, 21}, {768.5, 14.5}}
(-[SponsorCellNode layout]) (SponsorCellNode.m:189) Node {{0, 0}, {320, 63}}
Caption {{69, 21}, {487, 36}}
(-[SponsorCellNode layout]) (SponsorCellNode.m:189) Node {{0, 0}, {320, 63}}
Caption {{69, 21}, {714, 14.5}}

**Comment by appleguy** [comment]
@robmontesinos I actually mis-wrote the advice slightly.  Did you set .flexShrink on the text nodes?  Try setting it on both text nodes, but not the vertical stack spec.  If that doesn't work, add it to the vertical stack spec too.

Sorry I don't have the 100% right answer the first time around here, lots of stuff going on right now and using these Box Model patterns in practice has not been one of them!

**Comment by robmontesinos** [comment]
@appleguy  You are a Genius!! Here is what worked and a screenshot. I left the commented code in so you could see what I was trying.

```ObjectiveC

- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize {
    NSLog(@"captionSize: %@", NSStringFromCGSize(_captionSize));
    
    self.iconImageNode.preferredFrameSize = CGSizeMake(kSponsorIconSize, kSponsorIconSize);
//    self.titleTextNode.preferredFrameSize = CGSizeMake(_captionSize.width, kTitleHeight);
//    self.captionTextNode.preferredFrameSize = _captionSize;
    self.captionTextNode.flexShrink = YES;
    self.titleTextNode.flexShrink = YES;
   
    ASStackLayoutSpec *verticalStack = [[ASStackLayoutSpec alloc] init];
    verticalStack.direction = ASStackLayoutDirectionVertical;
    verticalStack.flexShrink = YES;
    [verticalStack setChildren:@[self.titleTextNode, self.captionTextNode]];
    
    ASStackLayoutSpec *horizontalStack = [[ASStackLayoutSpec alloc] init];
    horizontalStack.direction = ASStackLayoutDirectionHorizontal;
    horizontalStack.spacing = kInnerPadding;
    [horizontalStack setChildren:@[self.iconImageNode, verticalStack]];
    
    ASInsetLayoutSpec *insetSpec = [[ASInsetLayoutSpec alloc] init];
    insetSpec.insets = UIEdgeInsetsMake(kOuterPadding, kOuterPadding, kOuterPadding, kOuterPadding);
    insetSpec.child = horizontalStack;
    
    return insetSpec;
}

- (void)layout {
    [super layout];
    // Manually layout the divider.
    CGFloat pixelHeight = 1.0f / [[UIScreen mainScreen] scale];
    self.divider.frame = CGRectMake(0.0f, 0.0f, self.calculatedSize.width, pixelHeight);
    
//    self.captionTextNode.frame = CGRectMake(kTextNodeXPos, kTitleHeight + kOuterPadding, _captionSize.width, _captionSize.height);
    
    NSLog(@"%@\n%@", NSStringFromCGRect(self.frame), NSStringFromCGRect(self.captionTextNode.frame));
}


![Screenshot](https://dl.dropboxusercontent.com/u/47066165/ASDK-screenshot3.png)

**Comment by robmontesinos** [comment]
This was the easiest cell in terms of layout - at least, as they said in Saving Private Ryan, we're back in business. Thank you so much for your help.

**Comment by robmontesinos** [comment]
vertical stack flexShrink = YES was the ticket

**Comment by appleguy** [comment]
@robmontesinos just to confirm, you needed to set it on all three, right?  One framework improvement that would be nice — but may not be logically compatible with the overall layout rules of Flexbox... — would be if a stack containing all flexShrink = YES items were automatically flexShrink itself.

**Comment by robmontesinos** [comment]
Yes that is correct - all three - title, caption and vertical stack - code below

```ObjectiveC
- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize {
    self.iconImageNode.preferredFrameSize = CGSizeMake(kSponsorIconSize, kSponsorIconSize);
    self.captionTextNode.flexShrink = YES;
    self.titleTextNode.flexShrink = YES;
   
    ASStackLayoutSpec *verticalStack = [[ASStackLayoutSpec alloc] init];
    verticalStack.direction = ASStackLayoutDirectionVertical;
    verticalStack.flexShrink = YES;
    [verticalStack setChildren:@[self.titleTextNode, self.captionTextNode]];
    
    ASStackLayoutSpec *horizontalStack = [[ASStackLayoutSpec alloc] init];
    horizontalStack.direction = ASStackLayoutDirectionHorizontal;
    horizontalStack.spacing = kInnerPadding;
    [horizontalStack setChildren:@[self.iconImageNode, verticalStack]];
    
    ASInsetLayoutSpec *insetSpec = [[ASInsetLayoutSpec alloc] init];
    insetSpec.insets = UIEdgeInsetsMake(kOuterPadding, kOuterPadding, kOuterPadding, kOuterPadding);
    insetSpec.child = horizontalStack;
    
    return insetSpec;
}

**Comment by appleguy** [comment]
@robmontesinos I'm going to close out this task, but feel free to open another one for other issues with more complex cell layouts.  Thanks for showing the final code!

**Comment by robmontesinos** [comment]
You bet @appleguy  - 420 and out!!

**Comment by nguyenhuy** [comment]
@robmontesinos Do you somehow provide a fixed constrained height for each cell node? I asked this question because currently ASTableViewDataSource can't provide different constrained heights for each cells. And without a fixed constrained height (63, right?), the text nodes can grow vertically to wrap their texts. I think it makes sense to have -constrainedHeightForNodeAtIndexPath: in ASTableViewDataSource (note: not constrained size but constrained height, because we want table view cells to always fill the table view width). @appleguy, @rcancro: What do you think?

In case you are curious about why setting "all three" works, here is my explanation:
- Since the vertical stack is a child of a horizontal stack, enabling flexShrink on the vertical stack means it can shrink horizontally (flexShrink is effective on the stack direction). So the vertical stack fills the available width.
- The 2 text nodes are children of a vertical stack, so flexShrink means they can shrink vertically and thus fill the available height, which is 63 in this case.

Also, can the caption text (and title text) be really long? If so then it is possible that one of the 2 text nodes (likely titleTextNode because caption text is usually much longer) will be asked to shrink to a very small size that it cannot render even 1 line of text (or even asked to shrink to 0). This is because stack layout spec distributes stack-dimension violations evenly to all of its flexible children. Fortunately, you can easily handle this case. 
- Instead of enabling flexShrink on titleTextNode, you can set its maximumLineCount to 1 (or maybe 2). This change causes titleTextNode to always have enough space to render at most 1 line, and captionTextNode can fill the remaining height.
- Or if the caption is optional, you can just disable flexShrink on titleTextNode. This way, titleTextNode will take as much (vertical) space as it wants.

And of course, feel free to ask if you have problems with more complex layouts. We are happy to help :)

**Comment by robmontesinos** [comment]
@nguyenhuy I did provide a nodeSize to the ASCellNode as part of its init before I started with ASLayout, i.e. code below

```ObjectiveC
- (CGSize)calculateSizeThatFits:(CGSize)constrainedSize {
    return nodeSize;
}
```
Since @appleguy helped me get ASLayout to work, I noticed that my fixed height cell grew using the same node on another view controller. I was thinking of truncating the text but what you propose in terms of -constrainedHeightForNodeAtIndexPath: would be much better. I wonder if self.preferredFrameSize would work on the ASCellNode? I think it's a great idea @nguyenhuy

**Comment by nguyenhuy** [comment]
@appleguy: I think it's a good idea to have flexShrink enabled by default for text nodes. @rcancro: How about you?

> One framework improvement that would be nice — but may not be logically compatible with the overall layout rules of Flexbox... — would be if a stack containing all flexShrink = YES items were automatically flexShrink itself.

Since flexShrink is used against the stack dimension of the parent stack, I think this can confuse other developers. For example, if all children in a vertical stack can shrink (vertically), that doesn't always mean the vertical stack itself should shrink horizontally (if added inside a horizontally stack). It is true in @robmontesinos's layout because he happens to want a fixed height, but not in other (normal) cases.

**Comment by nguyenhuy** [comment]
@robmontesinos I don't think self.preferredFrameSize would work because -layoutSpecThatFits: would still be called ([code](https://github.com/facebook/AsyncDisplayKit/blob/master/AsyncDisplayKit/ASDisplayNode.mm#L1570). Assuming your layout code is implemented in the cell node.

-constrainedHeightForNodeAtIndexPath: is the best solution I can think of right now.

**Comment by robmontesinos** [comment]
@nguyenhuy When you're right, you're right. Thank you sir

**Comment by nguyenhuy** [comment]
You're welcome. It will take from 1 to 2 weeks for the new API to be implemented and reviewed (faster if you could contribute it!).

**Comment by robmontesinos** [comment]
That's fantastic @nguyenhuy - I will try to contribute in a few weeks - huge deadline on present project - thank you for your help in getting this ASLayout functionality going - I really like it

**Comment by nguyenhuy** [comment]
Sure thing. Good luck with your deadline :)
